--- joystick-20051019.orig/utils/Makefile
+++ joystick-20051019/utils/Makefile
@@ -41,8 +41,7 @@
 	$(RM) *.o *.swp $(PROGRAMS) *.orig *.rej map *~
 
 ffcfstress: ffcfstress.c
-	$(CC) -O2 -fno-implement-inlines -funsigned-char \
-	 -lm ffcfstress.c -o ffcfstress
+	$(CC) -O2 -funsigned-char -lm ffcfstress.c -o ffcfstress
 
 ffmvforce.o: ffmvforce.c
 	$(CC) -c $(CFLAGS) $(CPPFLAGS) $^ -o $@ `sdl-config --cflags`
--- joystick-20051019.orig/utils/ffmvforce.c
+++ joystick-20051019/utils/ffmvforce.c
@@ -147,7 +147,7 @@
 		fprintf(stderr, "Could not initialize SDL: %s\n", SDL_GetError());
 		exit(1);
 	}
-	on_exit(SDL_Quit, NULL);
+	atexit(SDL_Quit);
 	screen = SDL_SetVideoMode(WIN_W, WIN_H, 0, SDL_SWSURFACE);
 	if (screen == NULL) {
 		fprintf(stderr, "Could not set video mode: %s\n", SDL_GetError());
