Description: Pass flags required to enable hardened builds
Author: Stephen Kitt <steve@sk2.org>
Forwarded: yes

--- joystick.orig/utils/Makefile
+++ joystick/utils/Makefile
@@ -26,7 +26,7 @@
 #
 
 CC		= gcc
-CFLAGS		= -g -O2 -Wall -I../linux/include
+CPPFLAGS	+= -I../linux/include
 
 PROGRAMS	= inputattach jstest jscal fftest ffmvforce ffset \
 		  ffcfstress jscal-restore jscal-store
@@ -40,7 +40,7 @@
 	$(RM) *.o *.swp $(PROGRAMS) *.orig *.rej map *~
 
 ffcfstress: ffcfstress.c
-	$(CC) -O2 -funsigned-char ffcfstress.c -lm -o ffcfstress
+	$(CC) $(CFLAGS) $(CPPFLAGS) $(LDFLAGS) -funsigned-char ffcfstress.c -lm -o ffcfstress
 
 ffmvforce.o: ffmvforce.c
 	$(CC) -c $(CFLAGS) $(CPPFLAGS) $^ -o $@ `sdl-config --cflags`
@@ -53,14 +53,14 @@
 jscal.o: jscal.c axbtnmap.h
 
 jscal: jscal.o axbtnmap.o
-	$(CC) $(CFLAGS) $(CPPFLAGS) $^ -lm -o $@
+	$(CC) $(CFLAGS) $(CPPFLAGS) $(LDFLAGS) $^ -lm -o $@
 
 jstest.o: jstest.c axbtnmap.h
 
 jstest: jstest.o axbtnmap.o
 
 gencodes: gencodes.c scancodes.h
-	$(CC) $(CFLAGS) $(CPPFLAGS) gencodes.c -o gencodes
+	$(CC) $(CFLAGS) $(CPPFLAGS) $(LDFLAGS) gencodes.c -o gencodes
 
 jscal-restore: jscal-restore.in
 	sed "s^@@PREFIX@@^$(PREFIX)^g" < $^ > $@
